packages/wallet/src/KeyManagement/LedgerKeyAgent.ts

```diff
@@ -0,0 +1,217 @@
+import TransportWebHID from '@ledgerhq/hw-transport-webhid';
+import type Transport from '@ledgerhq/hw-transport';
+
+export interface LedgerKeyAgentProps {
+  networkId: Cardano.NetworkId;
+  accountIndex: number;
+  knownAddresses: GroupedAddress[];
+  extendedAccountPublicKey: Cardano.Bip32PublicKey;
+  deviceConnection?: AppAda;
+  communicationType: CommunicationType;
+}
+
+export interface CreateWithDevice {
+  networkId: Cardano.NetworkId;
+  accountIndex?: number;
+  communicationType: CommunicationType;
+}
+
+export interface GetXpubProps {
+  deviceConnection?: AppAda;
+  communicationType: CommunicationType;
+  accountIndex: number;
+}
+
+export interface CreateTransportProps {
+  communicationType: CommunicationType;
+  activeTransport?: TransportType;
+  devicePath?: string;
+}
+
+export class LedgerKeyAgent extends KeyAgentBase {
+  readonly #networkId: Cardano.NetworkId;
+  readonly #accountIndex: number;
+  readonly #knownAddresses: GroupedAddress[];
+  readonly #extendedAccountPublicKey: Cardano.Bip32PublicKey;
+  readonly #communicationType: CommunicationType;
+  readonly deviceConnection?: AppAda;
+
+  constructor({
+    networkId,
+    accountIndex,
+    knownAddresses,
+    extendedAccountPublicKey,
+    deviceConnection,
+    communicationType
+  }: LedgerKeyAgentProps) {
+    super();
+    this.#accountIndex = accountIndex;
+    this.#networkId = networkId;
+    this.#knownAddresses = knownAddresses;
+    this.#extendedAccountPublicKey = extendedAccountPublicKey;
+    this.#communicationType = communicationType;
+    this.deviceConnection = deviceConnection;
+  }
+
+  get networkId(): Cardano.NetworkId {
+    return this.#networkId;
+  }
+
+  get accountIndex(): number {
+    return this.#accountIndex;
+  }
+
+  get __typename(): KeyAgentType {
+    return KeyAgentType.Ledger;
+  }
+
+  get knownAddresses(): GroupedAddress[] {
+    return this.#knownAddresses;
+  }
+
+  get serializableData(): SerializableLedgerKeyAgentData {
+    return {
+      __typename: KeyAgentType.Ledger,
+      accountIndex: this.#accountIndex,
+      communicationType: this.#communicationType,
+      extendedAccountPublicKey: this.#extendedAccountPublicKey,
+      knownAddresses: this.#knownAddresses,
+      networkId: this.networkId
+    };
+  }
+
+  static async getHidDeviceList(): Promise<string[]> {
+    return await TransportNodeHid.list();
+  }
+
+  static async createTransport({
+    communicationType,
+    activeTransport,
+    devicePath = ''
+  }: CreateTransportProps): Promise<TransportType> {
+    if (communicationType === CommunicationType.Node) {
+      return await TransportNodeHid.open(devicePath);
+    }
+    return await (activeTransport && activeTransport instanceof TransportWebHID
+      ? TransportWebHID.open(activeTransport.device)
+      : TransportWebHID.request());
+  }
+
+  static async createDeviceConnection(activeTransport: Transport): Promise<AppAda> {
```

Comment:
Key agents would need to have different arguments then. Also we don't necessarily want to create a connection when creating the key agent - a connection is only needed for 1st load (getting xpub key) and signing.

Reply:
> need to have different arguments

Sure, but this can be typed generically.

> Also we don't necessarily want to create a connection when creating the key agent

Creating the key agent happens in the constructor, `initialize` keeps this separate. Is there another step I'm missing?